### PR TITLE
ZMI4/Py3: Merge with the fork of zms-publishing 

### PR DESCRIPTION
**Pull request by drfho**
HI @dataflake,
we did a lot of user tests with latest zope:master, but we did not programme any py-tests.
I think Products.LDAPUserFolder looks quite nice in it's new environment. 
Hope you like it, too
f

LUF_search0

### CODE REVIEW

**Comment by dataflake** [comment]
You guys need to fix the conficts first.

**Comment by dataflake** [comment]
To expand on that a little bit: There is no point in me taking a closer look until the PR is able to merge without conflict.

You probably want to merge `master` from here back into your own fork, then fix the conflicts. That will automatically fix the PR and I can take a look.

**Comment by drfho** [comment]
I suppose the the back-merge is done, now? After you confirm it, we will delete the fork, okay?

File: Products/LDAPUserFolder/LDAPDelegate.py [hunk 1]
```diff
@@ -568,7 +565,7 @@ def _clean_rdn(self, rdn):
         try:
             key, val = rdn.split('=')
             val = val.lstrip()
-            return '%s=%s' % (key, escape_dn_chars(val))
+            return '%s=%s' % (key, val)
```
Comment (dataflake) [inline Products/LDAPUserFolder/LDAPDelegate.py]: `escape_dn_chars` was called for a reason. Why did you take it out?

File: Products/LDAPUserFolder/LDAPUserFolder.py [hunk 2]
```diff
@@ -1234,7 +1237,7 @@ def getLDAPSchema(self):
     @security.protected(manage_users)
     def getSchemaDict(self):
         """ Retrieve schema as list of dictionaries """
-        all_items = sorted(self.getSchemaConfig().values())
+        all_items = self.getSchemaConfig().values()
```
Comment (dataflake) [inline Products/LDAPUserFolder/LDAPUserFolder.py]: Why did you remove the call to `sorted` here? If it failed bacause the return value of `values` is not a list then coerce it into a list. Leave the sorting.

Comment (drfho) [reply Products/LDAPUserFolder/LDAPUserFolder.py]: because sorted() generates  a TypeError:

TRACEBACK
"/home/dev/workspace/zopefoundation/Products.LDAPUserFolder/Products/LDAPUserFolder/LDAPUserFolder.py", line 1231, in getLDAPSchema
    raw_schema = self.getSchemaDict()
  File "/home/dev/workspace/zopefoundation/Products.LDAPUserFolder/Products/LDAPUserFolder/LDAPUserFolder.py", line 1240, in getSchemaDict
    all_items = sorted(self.getSchemaConfig().values())
TypeError: '<' not supported between instances of 'dict' and 'dict'

File: Products/LDAPUserFolder/LDAPDelegate.py (hunk 1 above)
Comment (drfho) [reply Products/LDAPUserFolder/LDAPDelegate.py]: Actually I don't know:  maybe it was (mis-) interpreted as a py2 residual and is was removed as a exclusion strategy (remove everyting what might cause errors)  - as a first step:
https://github.com/zms-publishing/Products.LDAPUserFolder/commit/9c97e82e6b5905e6bef311aee11a303049bda2e8#diff-f896aace4e37b3a85794c600852fa14d

I added the function again
https://github.com/zms-publishing/Products.LDAPUserFolder/commit/2f6f78add7212b360be65de3ca60bbfd806da491

File: Products/LDAPUserFolder/LDAPUserFolder.py (hunk 2 above)
Comment (dataflake) [reply Products/LDAPUserFolder/LDAPUserFolder.py]: Please replace the quick fix with something that reflects the intention:

`all_items = sorted(self.getSchemaConfig().values(), key=lambda x: x['ldap_name'])`

Comment (drfho) [reply Products/LDAPUserFolder/LDAPUserFolder.py]: done:
https://github.com/dataflake/Products.LDAPUserFolder/pull/3/commits/c45411be4b267bda997ded76e2c827e494d3cfc8
Took some time to guess the intention: you want to get the ldap schema representation  the administrator has configured in Zope sorted alphabetically by "LDAP Attribute Name" [ldap_names:list]. 
This definitly makes sense. 
Thank you.

Comment (dataflake) [reply Products/LDAPUserFolder/LDAPUserFolder.py]: That method is used (among other places) to get the schema items that show on the Schema tab. The sorting here is not su much about a specific item to sort on, just to make sure that the ordering is stable and doesn't change every time I click on the Schema tab.
